packages/sync/src/PersistentStore.ts

```diff
@@ -9,3 +8,199 @@ export interface PersistentStore<T> {
+  public storage: PersistentStore<PersistedData>;
+  private crypto = window.crypto;
+  private masterKey: CryptoKey | undefined;
+  private password: string;
+
+  constructor(storage: PersistentStore<PersistedData>, encryptionPassword: string) {
+    this.password = encryptionPassword;
+    this.storage = storage;
+    this.init();
+  }
+
+  /**
+   * Retrieves an encrypted item from storage.
+   * First extracts the initialisation vector that was previously prepended to the stored encrypted data
+   * @param key the key to use to retreive your value
+   */
+  public getItem(key: string) {
+    const retrieved = this.storage.getItem(key);
+    if (retrieved) {
+      if (this.masterKey) {
+        const cipherTextWithIv = retrieved.toString();
+        const cipherText = cipherTextWithIv.slice(cipherTextWithIv.indexOf(":") + 1);
+        const iv = this.strToArray(cipherTextWithIv.substr(0, cipherTextWithIv.indexOf(":")));
+        this.crypto.subtle.decrypt({ name: "AES-GCM", iv },
+          this.masterKey, this.strToArray(cipherText)).then(result => {
+            return this.arrayToStr(result);
+          });
+      }
+    }
+    return retrieved;
+  }
+
+  /**
+   * Stores an ecrypted item in storage. Generates a random initialisation vector for each encryption.
+   * This value is then prepended to the data and also stored for later decryption.
+   * @param key the key to use to store your value
+   * @param data the data to persist
+   */
+  public setItem(key: string, data: PersistedData) {
+    const iv = crypto.getRandomValues(new Uint8Array(12));
+    if (data) {
+      if (this.masterKey) {
+        this.crypto.subtle.encrypt({
+          name: "AES-GCM",
+          iv
+        },
+          this.masterKey,
+          this.strToArray(data.toString())
+        ).then(cipherText => {
+          const stringCipher = this.arrayToStr(cipherText);
+          this.storage.setItem(key, iv + ":" + stringCipher);
+        });
+      } else {
+        this.storage.setItem(key, data);
+      }
+    }
+  }
+
+  /**
+   * Remove an item from storage
+   * @param key the key to use to remove your value
+   */
+  public removeItem(key: string) {
+    this.storage.removeItem(key);
+  }
+
+  private async init() {
+    const masterKey = await this.storage.getItem("masterKey");
+    if (masterKey) {
+      const keyAsString = masterKey.toString();
+      const parsedKey = keyAsString.slice(keyAsString.indexOf(":") + 1);
+      await this.getSalt().then( async (res) => {
+        if (res) {
+          await this.unwrapCryptoKey(this.strToArray(parsedKey), res).then(unwrapped => unwrapped);
+        }
+      });
+    } else {
+      this.generateKey();
+    }
+  }
+
+  private generateKey() {
+    this.crypto.subtle.generateKey(
+      {
+        name: "AES-GCM",
+        length: 256
+      },
+      true,
+      ["encrypt", "decrypt"]
+    ).then(result => {
+      this.masterKey = result;
+      this.wrapCryptoKey(this.masterKey);
```

Comment:
this function does more than just generate the master key, it actually generates a key, and wrap it. So I suggest decouple the wrap part from this function. Chain them together in the init call to make the code easier to read.